tests/plugins/Qukeys/basic/test/testcase.cpp

```diff
@@ -0,0 +1,164 @@
+/* -*- mode: c++ -*-
+ * Copyright (C) 2020  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "testing/setup-googletest.h"
+
+#include "../sketch.h"
+
+SETUP_GOOGLETEST();
+
+namespace kaleidoscope {
+namespace testing {
+namespace {
+
+struct QukeyValues {
+  Key primary{Key_NoKey};
+  Key alternate{Key_NoKey};
+};
+constexpr KeyAddr key_addr_A{2, 1};
+constexpr KeyAddr key_addr_S{2, 2};
+constexpr KeyAddr key_addr_D{2, 3};
+constexpr KeyAddr key_addr_F{2, 4};
+constexpr KeyAddr key_addr_X{3, 2};
+
+using ::testing::IsEmpty;
+
+class QukeysBasic : public VirtualDeviceTest {};
+
+TEST_F(QukeysBasic, TapQukeyAlone) {
+
+  std::unique_ptr<State> state{nullptr};
+  std::set<uint8_t> expected_keycodes{};
```

Comment:
in terms of making the tests feel cleaner and clearer, making expected_keycodes something that can marshal Keys itself, would go a long way.

Reply:
I'm glad to hear that you're thinking the same thing as me. I think `expected_keycodes` would be a member variable structure, and would get populated by `Key` values via methods.